pages/design_documentation/ocf_io_engines.md

```diff
@@ -0,0 +1,269 @@
+---
+title: OCF I/O engines
+last_updated: August 17, 2022
+toc: false
+permalink: ocf_io_engines.html
+---
+
+To achieve a desired mode of caching, Open CAS Framework uses basic building
+blocks called IO Engines. These blocks implement specific algorithms, e.g.
+write-through or write-back and handle special (corner) cases as well.
+
+# OCF I/O engines - general description.
+OpenCAS can operate in [**several modes**](/cache_configuration.html), e.g.
+*write-back* or *write-through*. In general, OCF I/O engines are basic building
+blocks of a logic behind a particular cache mode. In other words, engines are
+sets of functions implementing the whole logic behind write and read operations
+for a given cache mode (and more, see below).
+
+This is an excerpt of a mapping table used in the OCF to call read or write
+handler of a particular engine (*write-through* and *write-back* engines 
+mappings are shown).
+
+~~~~~~~
+static const struct ocf_io_if IO_IFS[OCF_IO_PRIV_MAX_IF] = {
+	[OCF_IO_WT_IF] = {
+		.read = ocf_read_generic,       
+                .write = ocf_write_wt,          
+                .name = "Write Through"
+	}
+        [OCF_IO_WB_IF] = { 
+                .read = ocf_read_generic,       
+                .write = ocf_write_wb,          
+                .name = "Write Back"            
+        },
+[...]
+~~~~~~~
+
+For these particular engines, only the write path logic is different and
+specific engine functions are assigned accordingly. The read path is handled 
+by a generic read function.
+
+## I/O request handling by the engines
+
+When an I/O request is issued from an application using the OpenCAS software, 
+it is passed from the exported device (usually named /dev/casX-X) into the 
```

Comment:
We typically refer to it as **exported object**.
BTW do we want this documentation to be so OCL-centric? It describes OCF features, so making it more generic would help address also SPDK and potentially any other adapter.